cpp/src/arrow/compute/kernels/scalar_if_else_test.cc

```diff
@@ -624,6 +624,250 @@ TYPED_TEST(TestCaseWhenNumeric, ListOfType) {
+  auto values1_backing = DictArrayFromJSON(inner_type, "[0, null, 3, 1]", dict);
+  auto values2_backing = DictArrayFromJSON(inner_type, "[2, 1, null, 0]", dict);
+  auto values1 = make_list(ArrayFromJSON(int32(), "[0, 2, 2, 3, 4]"), values1_backing);
+  auto values2 = make_list(ArrayFromJSON(int32(), "[0, 1, 2, 2, 4]"), values2_backing);
+
+  CheckScalarNonRecursive(
+      "case_when", {MakeStruct({cond1, cond2}), values1, values2},
+      make_list(ArrayFromJSON(int32(), "[0, 2, 2, null, 2]"),
+                DictArrayFromJSON(inner_type, "[0, null]", R"(["a"])")));
+  CheckScalarNonRecursive(
+      "case_when",
+      {MakeStruct({cond1, cond2}), values1,
+       make_list(ArrayFromJSON(int32(), "[0, 1, null, 2, 4]"), values2_backing)},
+      make_list(ArrayFromJSON(int32(), "[0, 2, null, null, 2]"),
+                DictArrayFromJSON(inner_type, "[0, null]", R"(["a"])")));
+  CheckScalarNonRecursive(
+      "case_when",
+      {MakeStruct({cond1, cond2}), values1,
+       make_list(ArrayFromJSON(int32(), "[0, 1, null, 2, 4]"), values2_backing), values1},
+      make_list(ArrayFromJSON(int32(), "[0, 2, null, 2, 3]"),
+                DictArrayFromJSON(inner_type, "[0, null, 1]", R"(["a", "b"])")));
+
+  CheckScalarNonRecursive(
+      "case_when",
+      {
+          Datum(MakeStruct({cond1, cond2})),
+          Datum(std::make_shared<ListScalar>(
+              DictArrayFromJSON(inner_type, "[0, 1]", dict))),
+          Datum(std::make_shared<ListScalar>(
+              DictArrayFromJSON(inner_type, "[2, 3]", dict))),
+      },
+      make_list(ArrayFromJSON(int32(), "[0, 2, 4, null, 6]"),
+                DictArrayFromJSON(inner_type, "[0, 1, 0, 1, 2, 3]", dict)));
+
+  CheckScalarNonRecursive(
+      "case_when", {MakeStruct({Datum(true), Datum(false)}), values1, values2}, values1);
+  CheckScalarNonRecursive(
+      "case_when", {MakeStruct({Datum(false), Datum(true)}), values1, values2}, values2);
+  CheckScalarNonRecursive("case_when", {MakeStruct({Datum(false)}), values1, values2},
+                          values2);
+  CheckScalarNonRecursive("case_when",
+                          {MakeStruct({Datum(false), Datum(false)}), values1, values2},
+                          values_null);
+}
+
+TYPED_TEST(TestCaseWhenDict, DifferentDictionaries) {
+  auto type = dictionary(default_type_instance<TypeParam>(), utf8());
+  auto cond1 = ArrayFromJSON(boolean(), "[true, true, null, null]");
+  auto cond2 = ArrayFromJSON(boolean(), "[true, false, true, null]");
+  auto dict1 = R"(["a", null, "bc", "def"])";
+  auto dict2 = R"(["bc", "foo", null, "a"])";
+  auto dict3 = R"(["def", "a", "a", "bc"])";
+  auto values1_null = DictArrayFromJSON(type, "[null, null, null, null]", dict1);
+  auto values2_null = DictArrayFromJSON(type, "[null, null, null, null]", dict2);
+  auto values1 = DictArrayFromJSON(type, "[0, null, 3, 1]", dict1);
+  auto values2 = DictArrayFromJSON(type, "[2, 1, null, 0]", dict2);
```

Comment:
For some reason, it looks like the nulls in the indices are placed at the same indices as the nulls in the respective dictionaries.